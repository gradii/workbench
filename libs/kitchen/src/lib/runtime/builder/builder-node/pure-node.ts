import { isBoolean, isObject } from '@gradii/nanofn';
import { escapeHtml } from '../../helper';
import { isBinding, resolveBinding } from '../template-builder';

interface SimpleNode {
}

export type PureNode = Comment | Element | PureText | PureElementListNode | PureElementNode;


export class PureElementListNode implements SimpleNode {
  constructor(public elementNodes: PureNode[] = []) {
  }

  addElementNode(elementNode: PureElementNode) {
    this.elementNodes.push(elementNode);
  }
}

export class PureElementNode implements SimpleNode {
  parent: PureElementNode | null = null;

  public attributes: PureAttribute[] = [];

  public classAttribute        = [];
  public classBindingAttribute = [];
  public styleAttribute        = [];
  public styleBindingAttribute = [];

  textContent = null;

  constructor(public tagName: string,
              public isSlot: boolean      = false,
              public children: PureNode[] = []) {
  }

  addClass(value) {
    this.classAttribute.push(value);
    return this;
  }


  addClassBinding(key, value) {
    this.classBindingAttribute.push({ key, value });
    return this;
  }

  addStyle(value) {
    this.styleAttribute.push(value);
    return this;
  }

  addStyleBinding(key, value) {
    this.styleBindingAttribute.push({ key, value });
    return this;
  }

  addAttribute(name: string, value: string = null) {
    this.attributes.push(new PureAttribute(name, value));
    return this;
  }

  addInputAttribute(name: string, value: string | Record<string, string | any>) {
    this.attributes.push(new PureInputAttribute(name, value));
    return this;
  }

  addOutputAttribute(name: string, value: string | Record<string, string | any>) {
    this.attributes.push(new PureOutputAttribute(name, value));
    return this;
  }

  addTextContent(value: string) {
    this.textContent = value;
    return this;
  }
}

export class PureText implements SimpleNode {
  constructor(
    public value: string) {
  }
}

export class PureAttribute implements SimpleNode {
  constructor(
    public name: string, public value: string | Record<string, string | any>
  ) {
  }
}

export class PureClassAttribute extends PureAttribute {
  public readonly name = 'class';

  constructor(
    public value: string[]) {
    super('class', value);
  }
}

export class PureInputAttribute extends PureAttribute implements SimpleNode {
  constructor(
    public name: string, public value: string | Record<string, string | any>) {
    super(name, value);
  }
}

export class PureOutputAttribute extends PureAttribute implements SimpleNode {
  constructor(
    public name: string, public value: string | Record<string, string | any>) {
    super(name, value);
  }
}

//
// export class SimpleElement implements SimpleNode {
//   constructor(
//     public name: string, public attrs: PureAttribute[], public children: Node[]) {
//   }
// }

export class PureComment implements SimpleNode {
  constructor(public value: string | null) {
  }
}

function safeHtmlTagName(tagName: string) {
  return tagName.replace(/[^a-z0-9-]/gi, '');
}

function safeHtmlAttributeName(name: string) {
  return name.replace(/[^a-z0-9-_]/gi, '');
}

export class TemplateBuildCodeGenVisitor {
  visitNode(node: PureNode): string {
    if (node instanceof PureElementNode) {
      return this.visitElementNode(node);
    } else if (node instanceof PureElementListNode) {
      return this.visitElementListNode(node);
    }
    return '';
  }

  visitElementNode(elementNode: PureElementNode): string {
    const tagName = safeHtmlTagName(elementNode.tagName);
    let output    = `<${tagName}`;
    if (elementNode.classAttribute.length > 0) {
      const classAttributes = elementNode.classAttribute.map(value => safeHtmlAttributeName(value));
      output += ` class="${classAttributes.join(' ')}"`;
    }
    if (elementNode.classBindingAttribute.length > 0) {
      const classAttributes = elementNode.classBindingAttribute
        .map(value => {
          const _value = resolveBinding(value.value);
          return `'${safeHtmlAttributeName(value.key)}':${_value}`;
        });
      output += ` [ngClass]="{${classAttributes}}"`;
    }

    elementNode.attributes.forEach(attribute => {
      const attributeName = safeHtmlAttributeName(attribute.name);
      if (attribute instanceof PureInputAttribute) {
        if (isBinding(attribute.value)) {
          const value = resolveBinding(attribute.value);
          output += ` [${attributeName}]="${value}"`;
        } else if (isObject(attribute.value)) {
          const value = Object.keys(attribute.value).map(key => {
            const value = attribute.value[key];
            if (isBoolean(value)) {
              return `'${key}':${value}`;
            } else if (isBinding(value)) {
              return `'${key}':${resolveBinding(value)}`;
            } else {
              return `'${key}':'${attribute.value[key]}'`;
            }
          }).join(',');
          output += ` [${attributeName}]="{${value}}"`;
        }
      } else if (attribute instanceof PureOutputAttribute) {
        const value = resolveBinding(attribute.value);
        output += ` (${attributeName})="${value}"`;
      } else {
        if (attribute.value == null) {
          output += ` ${attributeName}`;
        } else {
          output += ` ${attributeName}="${attribute.value}"`;
        }
      }
    });
    output += '>';
    if (elementNode.textContent) {
      output += escapeHtml(elementNode.textContent);
    }
    if (elementNode.children.length) {
      output += elementNode.children.map(child => this.visitNode(child)).join('');
    }
    output += `</${tagName}>`;
    return output;
  }

  private visitElementListNode(node: PureElementListNode) {
    return node.elementNodes.map(elementNode => this.visitNode(elementNode)).join('');
  }
}
