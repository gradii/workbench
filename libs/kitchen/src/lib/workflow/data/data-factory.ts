import { KitchenComponentDataState } from '@common/public-api';
import stringifyObject from 'stringify-object';

export function dataPropsFactory(
  value: any,
  dataKey: string = 'data'
): { [key: string]: any; dataState: KitchenComponentDataState } {
  value = stringifyObject(value, { indent: '  ', singleQuotes: true, inlineCharacterLimit: 80 });
  return {
    dataState: {
      empty: true,
      error: false,
      actualValue: value,
      sampleValue: value,
      connected: false,
      dataKey
    },
    [dataKey]: value
  };
}
