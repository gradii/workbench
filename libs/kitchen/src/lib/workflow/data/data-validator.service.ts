import { Injectable } from '@angular/core';
import { KitchenComponent } from '@common/public-api';

import { NotValidError } from './exception';
import { DefinitionsService } from '../../definitions/definitions.service';

export type KitchenDataValidator = (data: any, component: KitchenComponent) => boolean;

@Injectable({ providedIn: 'root' })
export class DataValidator {
  constructor(private definitions: DefinitionsService) {
  }

  validate(component: KitchenComponent, data: any): null | string {
    const definition = this.definitions.getDefinition(component.definitionId);

    if (!definition.dataConsumer) {
      throw new Error('Data Validator: cannot validate component data as it is not marked as dataConsumer: true');
    }
    if (!definition.dataValidator) {
      throw new Error('Data Validator: dataValidator must be provided');
    }

    const validator = definition.dataValidator;
    let error = null;

    try {
      validator(data, component);
    } catch (e) {
      if (e instanceof NotValidError) {
        error = e.message;
      }
    }
    return error;
  }
}
