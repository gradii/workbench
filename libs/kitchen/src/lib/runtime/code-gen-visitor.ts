import { KitchenComponent, KitchenSlot } from '@common/public-api';
import { PureElementNode, PureNode } from './builder/builder-node/pure-node';
import { TemplateBuilder } from './builder/template-builder';
import { mixinSpaceFeature } from './feature-mixins/mixin-space-feature';
import { mixinAccordion } from './mixins/mixin-accordion';
import { mixinButton } from './mixins/mixin-button';
import { mixinCard } from './mixins/mixin-card';
import { mixinCheckbox } from './mixins/mixin-checkbox';
import { mixinContentSlot } from './mixins/mixin-content-slot';
import { mixinHeading } from './mixins/mixin-heading';
import { mixinIcon } from './mixins/mixin-icon';
import { mixinSlot } from './mixins/mixin-slot';
import { mixinSpace } from './mixins/mixin-space';

export class CodeGenVisitor extends mixinContentSlot(
  mixinSlot(
    mixinIcon(
      mixinAccordion(
        mixinCheckbox(
          mixinButton(
            mixinHeading(
              mixinCard(
                mixinSpace(
                  mixinSpaceFeature(
                    class {
                    }
                  )
                )
              )
            )
          )
        )
      )
    )
  )
) {

  ident            = 0;
  output: string[] = [];

  public templateBuilder = new TemplateBuilder();

  elementNode: PureElementNode;

  visitElement(node: KitchenComponent | KitchenSlot): PureElementNode {
    if (!Object.hasOwnProperty.call(node, 'definitionId')) {
      return this.visitElementSlot(node as KitchenSlot);
    } else {
      if ((node as KitchenComponent).definitionId === 'space') {
        return this.visitElementSpace(node as KitchenComponent);
      } else if ((node as KitchenComponent).definitionId === 'space-feature') {
        return this.visitElementSpaceFeature(node as KitchenComponent);
      } else if ((node as KitchenComponent).definitionId === 'card') {
        return this.visitElementCard(node as KitchenComponent);
      } else if ((node as KitchenComponent).definitionId === 'button') {
        return this.visitElementButton(node as KitchenComponent);
      } else if ((node as KitchenComponent).definitionId === 'checkbox') {
        return this.visitElementCheckbox(node as KitchenComponent);
      } else if ((node as KitchenComponent).definitionId === 'accordion') {
        return this.visitElementAccordion(node as KitchenComponent);
      } else if ((node as KitchenComponent).definitionId === 'heading') {
        return this.visitElementHeading(node as KitchenComponent);
      } else if ((node as KitchenComponent).definitionId === 'icon') {
        return this.visitElementIcon(node as KitchenComponent);
      } else {
        throw new Error('unknown element');
      }
    }
  }

  incIdent(level: number = 1) {
    this.ident += level;
  }

  decIdent(level: number = 1) {
    this.ident -= level;
  }

  write(content: string) {
    this.output.push(content);
  }

  writeLn(content: string) {
    this.output.push(`${content}`);
  }

  identWrite(content: string) {
    this.output.push(`${'  '.repeat(this.ident)}${content}`);
  }

  identWriteLn(content: string) {
    this.output.push(`${'  '.repeat(this.ident)}${content}\n`);
  }

  incWrite(content: string, ident?: number) {
    this.incIdent(ident);
    this.identWrite(content);
  }

  decWrite(content: string, ident?: number) {
    this.decIdent(ident);
    this.identWrite(content);
  }

  writeDec(content: string, ident?: number) {
    this.identWrite(content);
    this.decIdent(ident);
  }

  getOutput() {
    return this.output.join('');
  }

  flush() {
    const output = this.getOutput();
    this.output  = [];
    return output;
  }

  getElementOutput(elementNodes: PureNode[]){
    return this.templateBuilder.build(elementNodes);
  }
}