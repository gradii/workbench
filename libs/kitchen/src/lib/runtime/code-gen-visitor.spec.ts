import * as prettier from 'prettier';
import { PureElementNode, PureNode } from './builder/builder-node/pure-node';
import { CodeGenVisitor } from './code-gen-visitor';
import { dataTest } from './data.test';

describe('code-gen-visitor', () => {
  it('test generate code', () => {
    const codeGenVisitor        = new CodeGenVisitor();
    const elementNode: PureNode = codeGenVisitor.visitElement({
      componentList: [],
      'id'         : '1',
      'featureList': []
    });

    expect(elementNode.isSlot).toBe(true);
    expect(elementNode).toMatchSnapshot();
    const result = codeGenVisitor.getElementOutput([elementNode]);
    expect(result).toBe(
      `<div class="space direction-column" cdkScrollable uid="1" triDropFlexContainer id="1" triDropFlexContainerOrientation="vertical" kitchenRuntime kitchenRuntimeComponentId="1" kitchenRuntimeDefinitionId="slot"></div>`);
  });

  it('test generate code with element', () => {
    const codeGenVisitor                  = new CodeGenVisitor();
    const elementBuilder: PureElementNode = codeGenVisitor.visitElement(
      dataTest.content
    );


    let result = codeGenVisitor.getElementOutput([elementBuilder]);


    result = prettier.format(result, {
      parser: 'html'
    });
    expect(result).toMatchSnapshot();
  });

});