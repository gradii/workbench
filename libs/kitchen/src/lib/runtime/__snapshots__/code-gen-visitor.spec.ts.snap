// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code-gen-visitor test generate code 1`] = `
PureElementNode {
  "attributes": Array [
    PureAttribute {
      "name": "cdkScrollable",
      "value": null,
    },
    PureAttribute {
      "name": "uid",
      "value": "1",
    },
    PureAttribute {
      "name": "triDropFlexContainer",
      "value": null,
    },
    PureAttribute {
      "name": "id",
      "value": "1",
    },
    PureAttribute {
      "name": "triDropFlexContainerOrientation",
      "value": "vertical",
    },
    PureAttribute {
      "name": "_kitchenRuntime",
      "value": null,
    },
    PureAttribute {
      "name": "_kitchenRuntimeComponentId",
      "value": "1",
    },
    PureAttribute {
      "name": "_kitchenRuntimeDefinitionId",
      "value": "slot",
    },
  ],
  "classAttribute": Array [
    "space",
    "direction-column",
  ],
  "classBindingAttribute": Array [],
  "isSlot": true,
  "parent": null,
  "styleAttribute": Array [],
  "styleBindingAttribute": Array [],
  "tagName": "div",
  "textContent": null,
}
`;

exports[`code-gen-visitor test generate code with element 1`] = `
"<div
  class=\\"space direction-column\\"
  cdkScrollable
  uid=\\"1\\"
  triDropFlexContainer
  id=\\"1\\"
  triDropFlexContainerOrientation=\\"vertical\\"
  _kitchenRuntime
  _kitchenRuntimeComponentId=\\"1\\"
  _kitchenRuntimeDefinitionId=\\"slot\\"
>
  <div
    class=\\"direction-column\\"
    cdkScrollable
    uid=\\"15786492785660.41803516438866173\\"
    triDropFlexContainer
    id=\\"15786492785660.41803516438866173\\"
    _kitchenRuntime
    _kitchenRuntimeComponentId=\\"15786492785660.41803516438866173\\"
    _kitchenRuntimeDefinitionId=\\"space\\"
    triDrag
    [triDragData]=\\"{'id':'15786492785660.41803516438866173','name':'Space','parentSlotId':'1'}\\"
    class=\\"space\\"
    [triDropFlexContainerData]=\\"{'id':'15786492785660.41803516438866173','name':'Space','parentSlotId':'1'}\\"
    (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
  >
    <tri-card
      uid=\\"Card\\"
      _kitchenRuntime
      _kitchenRuntimeComponentId=\\"ryh1rJX3Gkk9Pg-Gr3Jeh\\"
      _kitchenRuntimeDefinitionId=\\"card\\"
      triDrag
      [triDragData]=\\"{'id':'ryh1rJX3Gkk9Pg-Gr3Jeh','name':'Card','parentSlotId':'15786492785660.736525773611667'}\\"
      [kitchenWithSize]=\\"{'widthValue':'100','widthUnit':'%','widthAuto':false,'heightValue':'300','heightUnit':'px','heightAuto':true}\\"
      ><tri-card-header
        class=\\"space direction-column\\"
        cdkScrollable
        uid=\\"UDFkg9xXOZyhpgRV-aK8S\\"
        triDropFlexContainer
        id=\\"UDFkg9xXOZyhpgRV-aK8S\\"
        triDropFlexContainerOrientation=\\"vertical\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"UDFkg9xXOZyhpgRV-aK8S\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        ><div
          class=\\"direction-row\\"
          cdkScrollable
          uid=\\"g2qrbvD2ZIkOfrXkjGhKm\\"
          triDropFlexContainer
          id=\\"g2qrbvD2ZIkOfrXkjGhKm\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"g2qrbvD2ZIkOfrXkjGhKm\\"
          _kitchenRuntimeDefinitionId=\\"space\\"
          triDrag
          [triDragData]=\\"{'id':'g2qrbvD2ZIkOfrXkjGhKm','name':'Space4','parentSlotId':'UDFkg9xXOZyhpgRV-aK8S'}\\"
          class=\\"space\\"
          [triDropFlexContainerData]=\\"{'id':'g2qrbvD2ZIkOfrXkjGhKm','name':'Space4','parentSlotId':'UDFkg9xXOZyhpgRV-aK8S'}\\"
          (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        >
          <h6
            uid=\\"Heading3\\"
            _kitchenRuntime
            _kitchenRuntimeComponentId=\\"UUdqcuIWMDbCG2wpxUk33\\"
            _kitchenRuntimeDefinitionId=\\"heading\\"
            triDrag
            [triDragData]=\\"{'id':'UUdqcuIWMDbCG2wpxUk33','name':'Heading3','parentSlotId':'JDSw0q7Eji5CLMfCw-3Ut'}\\"
          >
            Card Headerlll
          </h6>
        </div></tri-card-header
      ><tri-card-body
        [ngClass]=\\"{'flex-nowrap':'true','direction-row':'true','flex-nowrap':'true','direction-row':'true'}\\"
        triDropFlexContainer
        id=\\"v0iBzyWMoGZ532tdQ76dX\\"
        [triDropFlexContainerData]=\\"{'id':'LhJzILpx121ToMms8yK_Z','name':'space-feature','hostId':'v0iBzyWMoGZ532tdQ76dX','parentSlotId':'v0iBzyWMoGZ532tdQ76dX','slotId':'v0iBzyWMoGZ532tdQ76dX'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        cdkScrollable
        uid=\\"v0iBzyWMoGZ532tdQ76dX\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"v0iBzyWMoGZ532tdQ76dX\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        triDropFlexContainer
        id=\\"v0iBzyWMoGZ532tdQ76dX\\"
        [triDropFlexContainerData]=\\"{'id':'LhJzILpx121ToMms8yK_Z','name':'space-feature','hostId':'v0iBzyWMoGZ532tdQ76dX','parentSlotId':'v0iBzyWMoGZ532tdQ76dX','slotId':'v0iBzyWMoGZ532tdQ76dX'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        ><button
          triButton
          uid=\\"Button8\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"x1CEvk4V5TQMG5uId3BHg\\"
          _kitchenRuntimeDefinitionId=\\"button\\"
          triDrag
          triDragData=\\"[object Object]\\"
          color=\\"primary\\"
          variant=\\"fill\\"
          [disabled]=\\"false\\"
          icon=\\"star\\"
        >
          Button456</button
        ><button
          triButton
          uid=\\"Button7\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"006qEuawoT9mFRZMjhFfU\\"
          _kitchenRuntimeDefinitionId=\\"button\\"
          triDrag
          triDragData=\\"[object Object]\\"
          color=\\"primary\\"
          variant=\\"fill\\"
          [disabled]=\\"false\\"
          icon=\\"star\\"
        >
          Button789
        </button></tri-card-body
      ><tri-card-footer
        class=\\"space direction-column\\"
        cdkScrollable
        uid=\\"_i3qrx1AZMOFQj1pJVyTN\\"
        triDropFlexContainer
        id=\\"_i3qrx1AZMOFQj1pJVyTN\\"
        triDropFlexContainerOrientation=\\"vertical\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"_i3qrx1AZMOFQj1pJVyTN\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        ><div
          class=\\"direction-row\\"
          cdkScrollable
          uid=\\"iXC7O-MfrMQEh5vE3h6-9\\"
          triDropFlexContainer
          id=\\"iXC7O-MfrMQEh5vE3h6-9\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"iXC7O-MfrMQEh5vE3h6-9\\"
          _kitchenRuntimeDefinitionId=\\"space\\"
          triDrag
          [triDragData]=\\"{'id':'iXC7O-MfrMQEh5vE3h6-9','name':'Space6','parentSlotId':'_i3qrx1AZMOFQj1pJVyTN'}\\"
          class=\\"space\\"
          [triDropFlexContainerData]=\\"{'id':'iXC7O-MfrMQEh5vE3h6-9','name':'Space6','parentSlotId':'_i3qrx1AZMOFQj1pJVyTN'}\\"
          (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        >
          <div class=\\"empty\\"></div></div></tri-card-footer></tri-card
    ><tri-card
      uid=\\"Card2\\"
      _kitchenRuntime
      _kitchenRuntimeComponentId=\\"Gb63MmCEbha-cNDiTjp8a\\"
      _kitchenRuntimeDefinitionId=\\"card\\"
      triDrag
      [triDragData]=\\"{'id':'Gb63MmCEbha-cNDiTjp8a','name':'Card2','parentSlotId':'15786492785660.736525773611667'}\\"
      [kitchenWithSize]=\\"{'widthValue':'100','widthUnit':'%','widthAuto':false,'heightValue':'300','heightUnit':'px','heightAuto':true}\\"
      ><tri-card-header
        [ngClass]=\\"{'flex-nowrap':'true','direction-row':'true','flex-nowrap':'true','direction-row':'true'}\\"
        triDropFlexContainer
        id=\\"_G9bdhAQN-iUu1Zr5TWya\\"
        [triDropFlexContainerData]=\\"{'id':'6cMyMyIaI32dsldfPPHI2','name':'space-feature','hostId':'_G9bdhAQN-iUu1Zr5TWya','parentSlotId':'_G9bdhAQN-iUu1Zr5TWya','slotId':'_G9bdhAQN-iUu1Zr5TWya'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        cdkScrollable
        uid=\\"_G9bdhAQN-iUu1Zr5TWya\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"_G9bdhAQN-iUu1Zr5TWya\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        triDropFlexContainer
        id=\\"_G9bdhAQN-iUu1Zr5TWya\\"
        [triDropFlexContainerData]=\\"{'id':'6cMyMyIaI32dsldfPPHI2','name':'space-feature','hostId':'_G9bdhAQN-iUu1Zr5TWya','parentSlotId':'_G9bdhAQN-iUu1Zr5TWya','slotId':'_G9bdhAQN-iUu1Zr5TWya'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        ><h6
          uid=\\"Heading2\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"0i4zYZ_D2bFSxNlw3TnMK\\"
          _kitchenRuntimeDefinitionId=\\"heading\\"
          triDrag
          [triDragData]=\\"{'id':'0i4zYZ_D2bFSxNlw3TnMK','name':'Heading2','parentSlotId':'_G9bdhAQN-iUu1Zr5TWya'}\\"
        >
          Card Header
        </h6></tri-card-header
      ><tri-card-body
        [ngClass]=\\"{'flex-nowrap':'true','direction-column':'true','flex-nowrap':'true','direction-column':'true'}\\"
        triDropFlexContainer
        id=\\"S2QonCCC_-94t-aiuwBeB\\"
        [triDropFlexContainerData]=\\"{'id':'FFQDocQQajLsx2X2EuFw1','name':'space-feature','hostId':'S2QonCCC_-94t-aiuwBeB','parentSlotId':'S2QonCCC_-94t-aiuwBeB','slotId':'S2QonCCC_-94t-aiuwBeB'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        cdkScrollable
        uid=\\"S2QonCCC_-94t-aiuwBeB\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"S2QonCCC_-94t-aiuwBeB\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        triDropFlexContainer
        id=\\"S2QonCCC_-94t-aiuwBeB\\"
        [triDropFlexContainerData]=\\"{'id':'FFQDocQQajLsx2X2EuFw1','name':'space-feature','hostId':'S2QonCCC_-94t-aiuwBeB','parentSlotId':'S2QonCCC_-94t-aiuwBeB','slotId':'S2QonCCC_-94t-aiuwBeB'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        ><button
          triButton
          uid=\\"Button4\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"pNx3r_4IfjaMqnWpWnGF2\\"
          _kitchenRuntimeDefinitionId=\\"button\\"
          triDrag
          triDragData=\\"[object Object]\\"
          color=\\"primary\\"
          variant=\\"fill\\"
          [disabled]=\\"false\\"
          icon=\\"star\\"
        >
          Button999</button
        ><button
          triButton
          uid=\\"Button5\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"tkm50xPk9guNE3yw3H5b2\\"
          _kitchenRuntimeDefinitionId=\\"button\\"
          triDrag
          triDragData=\\"[object Object]\\"
          color=\\"primary\\"
          variant=\\"fill\\"
          [disabled]=\\"false\\"
          icon=\\"star\\"
        >
          Buttonzzz
        </button></tri-card-body
      ><tri-card-footer
        [ngClass]=\\"{'flex-nowrap':'true','direction-row':'true','flex-nowrap':'true','direction-row':'true'}\\"
        triDropFlexContainer
        id=\\"Oys8q1_-aQ-zIeihcuhiV\\"
        [triDropFlexContainerData]=\\"{'id':'07zOrXvbNGK4w0o2WK1ne','name':'space-feature','hostId':'Oys8q1_-aQ-zIeihcuhiV','parentSlotId':'Oys8q1_-aQ-zIeihcuhiV','slotId':'Oys8q1_-aQ-zIeihcuhiV'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        cdkScrollable
        uid=\\"Oys8q1_-aQ-zIeihcuhiV\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"Oys8q1_-aQ-zIeihcuhiV\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        triDropFlexContainer
        id=\\"Oys8q1_-aQ-zIeihcuhiV\\"
        [triDropFlexContainerData]=\\"{'id':'07zOrXvbNGK4w0o2WK1ne','name':'space-feature','hostId':'Oys8q1_-aQ-zIeihcuhiV','parentSlotId':'Oys8q1_-aQ-zIeihcuhiV','slotId':'Oys8q1_-aQ-zIeihcuhiV'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        ><button
          triButton
          uid=\\"Button3\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"TZshbru7R-5lZOtYAiGas\\"
          _kitchenRuntimeDefinitionId=\\"button\\"
          triDrag
          triDragData=\\"[object Object]\\"
          color=\\"danger\\"
          variant=\\"fill\\"
          [disabled]=\\"false\\"
          icon=\\"star\\"
        >
          Buttondg</button
        ><button
          triButton
          uid=\\"Button1\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"WMT6kYoT_IEmNdi5eKy7c\\"
          _kitchenRuntimeDefinitionId=\\"button\\"
          triDrag
          triDragData=\\"[object Object]\\"
          color=\\"warning\\"
          variant=\\"fill\\"
          [disabled]=\\"false\\"
          icon=\\"star\\"
        >
          Button2345
        </button></tri-card-footer
      ></tri-card
    ><tri-card
      uid=\\"Card3\\"
      _kitchenRuntime
      _kitchenRuntimeComponentId=\\"FoflPrF9JT4OTg_RNErdT\\"
      _kitchenRuntimeDefinitionId=\\"card\\"
      triDrag
      [triDragData]=\\"{'id':'FoflPrF9JT4OTg_RNErdT','name':'Card3','parentSlotId':'15786492785660.736525773611667'}\\"
      [kitchenWithSize]=\\"{'widthValue':'100','widthUnit':'%','widthAuto':false,'heightValue':'300','heightUnit':'px','heightAuto':true}\\"
      ><tri-card-header
        [ngClass]=\\"{'flex-nowrap':'true','direction-row':'true','flex-nowrap':'true','direction-row':'true'}\\"
        triDropFlexContainer
        id=\\"cwDyXmOciWxyD0-JmWCxp\\"
        [triDropFlexContainerData]=\\"{'id':'qjteET_qkTLh512JlMLGk','name':'space-feature','hostId':'cwDyXmOciWxyD0-JmWCxp','parentSlotId':'cwDyXmOciWxyD0-JmWCxp','slotId':'cwDyXmOciWxyD0-JmWCxp'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        cdkScrollable
        uid=\\"cwDyXmOciWxyD0-JmWCxp\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"cwDyXmOciWxyD0-JmWCxp\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        triDropFlexContainer
        id=\\"cwDyXmOciWxyD0-JmWCxp\\"
        [triDropFlexContainerData]=\\"{'id':'qjteET_qkTLh512JlMLGk','name':'space-feature','hostId':'cwDyXmOciWxyD0-JmWCxp','parentSlotId':'cwDyXmOciWxyD0-JmWCxp','slotId':'cwDyXmOciWxyD0-JmWCxp'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
      ></tri-card-header
      ><tri-card-body
        [ngClass]=\\"{'flex-nowrap':'true','direction-row':'true','flex-nowrap':'true','direction-row':'true'}\\"
        triDropFlexContainer
        id=\\"FxKLEFS_f-pk-U60Mro32\\"
        [triDropFlexContainerData]=\\"{'id':'LhJzILpx121ToMms8yU_I','name':'space-feature','hostId':'FxKLEFS_f-pk-U60Mro32','parentSlotId':'FxKLEFS_f-pk-U60Mro32','slotId':'FxKLEFS_f-pk-U60Mro32'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        cdkScrollable
        uid=\\"FxKLEFS_f-pk-U60Mro32\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"FxKLEFS_f-pk-U60Mro32\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        triDropFlexContainer
        id=\\"FxKLEFS_f-pk-U60Mro32\\"
        [triDropFlexContainerData]=\\"{'id':'LhJzILpx121ToMms8yU_I','name':'space-feature','hostId':'FxKLEFS_f-pk-U60Mro32','parentSlotId':'FxKLEFS_f-pk-U60Mro32','slotId':'FxKLEFS_f-pk-U60Mro32'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
      ></tri-card-body
      ><tri-card-footer
        [ngClass]=\\"{'flex-nowrap':'true','direction-row':'true','flex-nowrap':'true','direction-row':'true'}\\"
        triDropFlexContainer
        id=\\"Qm6H0N7R31W7sNHMoARi5\\"
        [triDropFlexContainerData]=\\"{'id':'e_G8eg87o5qYyc--o0ec0','name':'space-feature','hostId':'Qm6H0N7R31W7sNHMoARi5','parentSlotId':'Qm6H0N7R31W7sNHMoARi5','slotId':'Qm6H0N7R31W7sNHMoARi5'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        cdkScrollable
        uid=\\"Qm6H0N7R31W7sNHMoARi5\\"
        _kitchenRuntime
        _kitchenRuntimeComponentId=\\"Qm6H0N7R31W7sNHMoARi5\\"
        _kitchenRuntimeDefinitionId=\\"slot\\"
        triDropFlexContainer
        id=\\"Qm6H0N7R31W7sNHMoARi5\\"
        [triDropFlexContainerData]=\\"{'id':'e_G8eg87o5qYyc--o0ec0','name':'space-feature','hostId':'Qm6H0N7R31W7sNHMoARi5','parentSlotId':'Qm6H0N7R31W7sNHMoARi5','slotId':'Qm6H0N7R31W7sNHMoARi5'}\\"
        (triDropFlexContainerDropped)=\\"onDragDropped($event)\\"
        ><button
          triButton
          uid=\\"Button\\"
          _kitchenRuntime
          _kitchenRuntimeComponentId=\\"0gYcXB0cZOutRIlfsOt31\\"
          _kitchenRuntimeDefinitionId=\\"button\\"
          triDrag
          triDragData=\\"[object Object]\\"
          color=\\"info\\"
          variant=\\"fill\\"
          [disabled]=\\"false\\"
          icon=\\"star\\"
        >
          Button
        </button></tri-card-footer
      ></tri-card
    ><tri-accordion
      uid=\\"Accordion\\"
      triDrag
      [triDragData]=\\"{'id':'hSqKZG_y_hB_r3E7-G6CD','name':'Accordion','parentSlotId':'15786492785660.736525773611667'}\\"
      _kitchenRuntime
      _kitchenRuntimeComponentId=\\"hSqKZG_y_hB_r3E7-G6CD\\"
      _kitchenRuntimeDefinitionId=\\"accordion\\"
    ></tri-accordion>
  </div>
</div>
"
`;
